### ISSUE TITLE
Bot does not stop talking

### ISSUE THREAD

**Issue by Kwisss** [jmoney7823956789378/oobabot#13]
I've added the <|eot_id|> stop token but it does not work?

**Comment by seaweeduk** [comment]
Same issue, also tried <|eot_id|> but it just goes on and on inventing replies from other people

**Comment by seaweeduk** [comment]
New to this but with these settings it seems to work somewhat ok with the stock llama3 model without continuing the conversation for other people but I don't really know what I'm doing... 

```
  # The BOS token that should be inserted before the system block.
  system_sequence_prefix: "<|begin_of_text|><|start_header_id|>system<|end_header_id|>\n"
  # The EOS token that should be inserted after the system block.
  system_sequence_suffix: "<|eot_id|><|start_header_id|>user<|end_header_id|>"
  # The BOS token that should be inserted before the user message block.
  user_sequence_prefix: ""
  # The EOS token that should be inserted after the user message block.
  user_sequence_suffix: "<|eot_id|>"
  # The BOS token that should be inserted before the bot message block.
  bot_sequence_prefix: "<|start_header_id|>assistant<|end_header_id|>\n\n"
  # The EOS token that should be inserted after the bot message block.
  bot_sequence_suffix: "<|eot_id|>"
```

**Comment by AlanMW** [comment]
Make sure you are using an instruct model and not a chat or base model.

**Comment by seaweeduk** [comment]
Yeah I'm using it with an instruct model, seems to work pretty good now I figured the template out.

Only issue I'm having is when certain characters appear it seems to drop the input. For example when prompting for image generation with a lora in the prompt with the format  ```<lora:Loraname:1>``` We also tried teaching it to say discord emotes but they also seem to get dropped, probably for a similar reason. I'm still digging through the code to try and understand it better...

**Comment by xBelladonna** [comment]
I am the one who implemented the instruct prompt templating system and I feel like I didn't adequately clarify some things.

Firstly it should be noted that there is a big difference between "stop markers" in the discord settings section, and the "stop" parameter under the oobabooga request parameters. The stop markers *do not stop generation* but instead are used by the immersion breaking filter, which allows generation to continue, while filtering out immersion breaking sections based on heuristics that the stop markers are used in. Any entries under the oobabooga "stop" request param will be added to the stop markers automatically, but not the other way around.

Secondly, the instruct template system will automatically add any sequence suffixes to the "stop" request param, so that generation ceases when these are encountered. If you add your end of turn token there, it should stop generation, but if you are only adding them as stop markers under discord settings, generation will not stop. I'll try to adequately explain this in my next release coming up shortly on my own fork.

Also with regards to the special characters issue in image prompts, I know why that's happening and I have partially fixed the issue in my own fork. It hasn't been a priority so far but I'll try to include a full fix for it in the next release. The issue simply comes down to a regex pattern that doesn't adequately cover the range of characters that might be present in a typical stable diffusion prompt, and you can find the relevant lines [here](https://github.com/xBelladonna/oobabot/blob/937086e1f868ac7d98770d70d87844cb0cedbdb7/src/oobabot/image_generator.py#L286)

**Comment by Kwisss** [comment]
@xBelladonna 
I have been busy these past months, forked out some dirty solutions for myself in the main time. 
But I will check out your fork out these couple of days, could you open a discussion or a issues tab for some collaboration on some issues?

**Comment by xBelladonna** [comment]
> @xBelladonna I have been busy these past months, forked out some dirty solutions for myself in the main time. But I will check out your fork out these couple of days, could you open a discussion or a issues tab for some collaboration on some issues?

@Kwisss I know what that can be like. I've been fairly busy myself, but in any case I have enabled Issues and Discussions on my repository.

Please be aware of two things though:
- I have made some serious changes under the hood (currently only in the `develop` branch, not `main` or any tagged releases) that change a lot about how the bot works, and introduces several breaking changes requiring some manual migration of configuration. I plan on overhauling everything much more significantly again, with the addition of several databases for more robust and dynamic (at-runtime) configuration, and long-term memory support.
- My fork has ended up evolving into a highly personal project that I mostly maintain for myself and my own use. I publish the code publicly for any case where others might find it useful, but my priority in terms of the development of my fork is simply "what I want and need". I therefore don't have any plans for maintaining this as a community project, publishing to PyPI, etc. Having said that, I am open to suggestions and collaboration, should it fit with my goals and desires. In any case, I welcome opening issues for bugs that I haven't caught so I can improve the quality of the code for both myself and others.
